mcp/tools/tools_coder.go

```diff
@@ -0,0 +1,381 @@
+		}
+
+		var buf bytes.Buffer
+		if err := json.NewEncoder(&buf).Encode(me); err != nil {
+			return nil, xerrors.Errorf("Failed to encode the current user: %s", err.Error())
+		}
+
+		return &mcp.CallToolResult{
+			Content: []mcp.Content{
+				mcp.NewTextContent(strings.TrimSpace(buf.String())),
+			},
+		}, nil
+	}
+}
+
+// Example payload:
+// {"jsonrpc":"2.0","id":1,"method":"tools/call", "params": {"name": "coder_list_workspaces", "arguments": {"owner": "me", "offset": 0, "limit": 10}}}
+func handleCoderListWorkspaces(deps ToolDeps) mcpserver.ToolHandlerFunc {
+	return func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		if deps.Client == nil {
+			return nil, xerrors.New("developer error: client is required")
+		}
+		args := request.Params.Arguments
+
+		owner, ok := args["owner"].(string)
+		if !ok {
+			owner = codersdk.Me
+		}
+
+		offset, ok := args["offset"].(int)
+		if !ok || offset < 0 {
+			offset = 0
+		}
+		limit, ok := args["limit"].(int)
+		if !ok || limit <= 0 {
+			limit = 10
+		}
+
+		workspaces, err := deps.Client.Workspaces(ctx, codersdk.WorkspaceFilter{
+			Owner:  owner,
+			Offset: offset,
+			Limit:  limit,
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("failed to fetch workspaces: %w", err)
+		}
+
+		// Encode it as JSON. TODO: It might be nicer for the agent to have a tabulated response.
+		data, err := json.Marshal(workspaces)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to encode workspaces: %s", err.Error())
+		}
+
+		return &mcp.CallToolResult{
+			Content: []mcp.Content{
+				mcp.NewTextContent(string(data)),
+			},
+		}, nil
+	}
+}
+
+// Example payload:
+// {"jsonrpc":"2.0","id":1,"method":"tools/call", "params": {"name": "coder_get_workspace", "arguments": {"workspace": "dev"}}}
+func handleCoderGetWorkspace(deps ToolDeps) mcpserver.ToolHandlerFunc {
+	return func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
+		if deps.Client == nil {
+			return nil, xerrors.New("developer error: client is required")
+		}
+		args := request.Params.Arguments
+
+		wsArg, ok := args["workspace"].(string)
+		if !ok {
+			return nil, xerrors.New("workspace is required")
+		}
+
+		workspace, err := getWorkspaceByIDOrOwnerName(ctx, deps.Client, wsArg)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to fetch workspace: %w", err)
+		}
+
+		workspaceJSON, err := json.Marshal(workspace)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to encode workspace: %w", err)
+		}
+
+		return &mcp.CallToolResult{
+			Content: []mcp.Content{
+				mcp.NewTextContent(string(workspaceJSON)),
+			},
+		}, nil
+	}
+}
+
+// Example payload:
+// {"jsonrpc":"2.0","id":1,"method":"tools/call", "params": {"name": "coder_workspace_exec", "arguments": {"workspace": "dev", "command": "ps -ef"}}}
+func handleCoderWorkspaceExec(deps ToolDeps) mcpserver.ToolHandlerFunc {
```

Comment:
Is there any way to pass tracing data? for instance, for how long was the command running?

Reply:
For whom is this data intended? For now, I've returned it in the response as JSON.

Reply:
I was wondering if the MCP protocol has space for such metadata. If somebody wants to visualize some metrics around calls, it seems to be channel to pass all tracing info.

Reply:
There doesn't seem to be, from what I can tell.